### PR TITLE
refactor(pool): implement resource-like pattern

### PR DESCRIPTION
**Pull request by nicojs**
Implement resource like pattern. Schedule work to be done on the pool with `pool.schedule(input$, task)`.

The pool itself is responsible to make sure only one task is executed on one worker at any one time. The pool is also responsible to recycle the worker after it is done with a task.

### CODE REVIEW

**Comment by nicojs** [comment]
Thanks for the help in thinking this out @hugo-vrijswijk! As a token of appreciation, you may now ... 🥁 review this PR! 🍰

**Comment by hugo-vrijswijk** [comment]
Tests are red. 😬

**Comment by nicojs** [comment]
Green, finally 💚

**Comment by nicojs** [comment]
I very much like this improvement. I always found the `Pool` api to have a broken abstraction. This is a neat solution

File: packages/core/src/concurrent/pool.ts [hunk 1]
```diff
@@ -25,27 +25,23 @@ export function createCheckerPool(factory: () => Checker, concurrencyToken$: Obs
...
+export class Pool<TWorker extends Worker> implements Disposable {
+  private readonly createdWorkers: TWorker[] = [];
+  private readonly worker$: Observable<TWorker>;
 
-  constructor(factory: () => T, concurrencyToken$: Observable<number>) {
-    this.createdWorker$ = concurrencyToken$.pipe(
-      flatMap(async () => {
+  constructor(factory: () => TWorker, concurrencyToken$: Observable<number>) {
+    this.worker$ = concurrencyToken$.pipe(
+      mergeMap(async () => {
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/core/src/concurrent/pool.ts:39]: `flatMap` is deprecated.

https://github.com/ReactiveX/rxjs/blob/7bbd37f53397193cf1371b93c1f93b18c071474f/src/internal/operators/mergeMap.ts#L97-L100

File: packages/core/src/concurrent/pool.ts [hunk 2]
```diff
@@ -55,23 +51,34 @@ export class Pool<T extends Worker> implements Disposable {
...
-    this.recycleBin.next(worker);
+  public schedule<TIn, TOut>(input$: Observable<TIn>, task: (worker: TWorker, input: TIn) => Promise<TOut> | TOut): Observable<TOut> {
+    const recycleBin = new Subject<TWorker>();
+    const worker$ = merge(recycleBin, this.worker$);
+
+    return zip(worker$, input$).pipe(
+      mergeMap(async ([worker, input]) => {
+        const output = await task(worker, input);
+        //  Recycles a worker so its re-emitted from the `worker$` observable.
+        recycleBin.next(worker);
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/core/src/concurrent/pool.ts:77]: When the task throws an error, the `error` callback is executed and the entire stream is closed see https://rxjs.dev/guide/observer. Not even `complete` will be executed and the recycle bin will never be completed.

I don't think that is a problem. The completing of `recycleBin` here might not be needed at all. What do you think? I find the recourses scares on this.

**Pull request merged**